clippy_lints/src/call_missing_target_feature.rs

```diff
@@ -0,0 +1,190 @@
+                CALL_MISSING_TARGET_FEATURE,
+                expr.span,
+                "this call requires target features that the surrounding function does not enable",
+                |diag| {
+                    diag.span_label(
+                        expr.span,
+                        "this function call requires target features to be enabled".to_string(),
+                    );
+
+                    if let Some(caller_item) = caller_item(cx, expr) {
+                        let hir::ItemKind::Fn(fn_sig, _, _) = caller_item.kind else {
+                            unreachable!()
+                        };
+
+                        let mut suggestions = Vec::with_capacity(2);
+
+                        let Some(indent) = indentation(cx, caller_item.span) else {
+                            return;
+                        };
+
+                        let lo_span = caller_item.span.with_hi(caller_item.span.lo());
+
+                        if let hir::Safety::Safe = fn_sig.header.safety {
+                            if caller_item.vis_span.is_empty() {
+                                suggestions.push((lo_span, format!("{attr}\n{indent}unsafe ")));
+                            } else {
+                                suggestions.push((lo_span, format!("{attr}\n{indent}")));
+                                suggestions.push((caller_item.vis_span.shrink_to_hi(), " unsafe".to_string()));
+                            }
+                        }
+
+                        diag.multipart_suggestion_verbose(
+                            "add the missing target features to the surrounding function",
+                            suggestions,
+                            rustc_errors::Applicability::MaybeIncorrect,
+                        );
+                    }
+                },
+            );
+        }
+    }
+}
+
+fn callee_def_id(cx: &LateContext<'_>, expr: &hir::Expr<'_>) -> Option<DefId> {
+    match expr.kind {
+        hir::ExprKind::Call(path, _) => {
+            if let hir::ExprKind::Path(ref qpath) = path.kind
+                && let Res::Def(_, did) = cx.qpath_res(qpath, path.hir_id)
+            {
+                Some(did)
+            } else {
+                None
+            }
+        },
+        hir::ExprKind::MethodCall(..) => cx.typeck_results().type_dependent_def_id(expr.hir_id),
+        _ => None,
+    }
+}
+
+fn caller_def_id<'tcx>(cx: &LateContext<'tcx>, expr: &hir::Expr<'tcx>) -> Option<DefId> {
+    let item = caller_item(cx, expr)?;
+    Some(item.owner_id.to_def_id())
+}
+
+fn caller_item<'tcx>(cx: &LateContext<'tcx>, expr: &hir::Expr<'tcx>) -> Option<&'tcx hir::Item<'tcx>> {
+    for (_hir_id, node) in cx.tcx.hir().parent_iter(expr.hir_id) {
+        if let hir::Node::Item(
+            item @ hir::Item {
+                kind: hir::ItemKind::Fn(..),
+                ..
+            },
+        ) = node
+        {
+            return Some(item);
+        }
+    }
+
+    None
+}
```

Comment:
If you switch to `codegen_fn_attrs` then you will no longer need to walk up the HIR tree. Getting the `DefId` for `LateContext::enclosing_body` will be enough.